doc/threshold/SONiC Threshold feature spec.md

```diff
@@ -0,0 +1,589 @@
+		  * [3.6.4 Clear Commands](#3_6_4-clear-commands)
+		  * [3.6.5 Debug Commands](#3_6_5-debug-commands)
+		  * [3.6.6 REST API Support](#3_6_6-rest-api-support)  
+  * [4 Flow Diagrams](#4-flow-diagrams)
+	  * [4.1 Config Call Flow](#4_1-config-call-flow)
+	  * [4.2 Breach Event Call Flow](#4_2-breach-event-call-flow)
+  * [5 Error Handling](#5-error-handling)
+  * [6 Serviceability And Debug](#6-serviceability-and-debug)
+  * [7 Warm Boot Support](#7-warm-boot-support)
+  * [8 Scalability](#8-scalability)
+  * [9 Unit Test](#9-unit-test)
+  * [10 Internal Design Information](#10-internal-design-information)
+
+# List of Tables
+[Table 1: Abbreviations](#table-1-abbreviations)
+
+# Revision
+| Rev |     Date    |       Author       | Change Description                |
+|:---:|:-----------:|:------------------:|-----------------------------------|
+| 0.1 | 06/12/2019  |   Shirisha Dasari  | Initial version                   |
+
+# About this Manual
+This document provides general information about the threshold feature implementation in SONiC.
+
+# Scope
+This document describes the high level design of threshold feature. 
+
+# Definition/Abbreviation
+### Table 1: Abbreviations
+| **Term**                 | **Meaning**                         |
+|--------------------------|-------------------------------------|
+| SAI                      | Switch abstraction interface        |
+| TAM                      | Telemetry and monitoring            |
+| Watermark                | Peak counter value reached          |
+| ThresholdMgr             | Threshold manager                   |
+
+# 1 Feature Overview
+The threshold feature allows configuration of a threshold on supported buffers in ingress and egress. A threshold breach notification (entry update in COUNTERS\_DB) is generated on breach of the configured buffer threshold in ASIC.   
+
+## 1.1 Requirements
+### 1.1.1 Functional Requirements
+
+1.0 Threshold feature allows the user to configure a threshold on a buffer (i.e buffer usage) and record all threshold breach events for that buffer.
+
+1.1 On breach of the configured threshold in ASIC, a breach event entry is added to COUNTERS\_DB. User can list out the breach event data through a CLI command. In addition, the breach entries in COUNTERS\_DB are also available to SONiC telemetry feature as well.
+
+1.2 The threshold feature can be used to track and be notified of congestion events in the network real-time.
+
+1.3 Monitoring of buffer usage can be achieved via the watermark feature available in SONiC.
+
+1.4 Along with the watermark feature, the threshold feature provides data that can be used to design applications to monitor MMU buffer usage.  
+
+2.0.0 The threshold for a buffer must be configured in terms of percentage of buffer allocation.
+ 
+2.0.1 The configured threshold can be cleared at any point by the user.
+
+2.0.2 The supported counters for threshold feature are listed below.
+
```

Comment:
Pls add a switch capability check to get the list of counters for which threshold feature is supported.

Reply:
Hi,
I'll explore this and get back to you on the same.

Reply:
Hi,

A new switch capability for list of counters would be a new requirement on SAI spec. It will need to be proposed and followed up with the SAI community.